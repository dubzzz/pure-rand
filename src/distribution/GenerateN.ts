import type { RandomGenerator } from '../types/RandomGenerator';
import { unsafeGenerateN } from './UnsafeGenerateN';

export function generateN(rng: RandomGenerator, num: number): [number[], RandomGenerator] {
  const nextRng = rng.clone();
  const out = unsafeGenerateN(nextRng, num);
  return [out, nextRng];
}
