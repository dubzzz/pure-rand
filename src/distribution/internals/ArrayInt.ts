/**
 * An ArrayInt represents an integer larger than what can be represented in classical JavaScript.
 * The values stored in data must be in the range [0, 0xffffffff].
 *
 * @example
 * ```js
 * { sign:  1, data: [ 42 ] } // = 42
 * { sign: -1, data: [ 42 ] } // = -42
 * { sign: -1, data: [ 5, 42 ] } // = -1 * (5 * 2**32 + 42)
 * { sign: -1, data: [ 1, 5, 42 ] } // = -1 * (1 * 2**64 + 5 * 2**32 + 42)
 * ```
 */
export type ArrayInt = {
  /**
   * Sign of the represented number
   */
  sign: -1 | 1;
  /**
   * Value of the number, must only contain numbers in the range [0, 0xffffffff]
   */
  data: number[];
};

/**
 * Add two ArrayInt
 * @internal
 */
export function addArrayIntToNew(arrayIntA: ArrayInt, arrayIntB: ArrayInt): ArrayInt {
  if (arrayIntA.sign !== arrayIntB.sign) {
    return substractArrayIntToNew(arrayIntA, { sign: -arrayIntB.sign as -1 | 1, data: arrayIntB.data });
  }
  const data: number[] = [];
  let reminder = 0;
  const dataA = arrayIntA.data;
  const dataB = arrayIntB.data;
  for (let indexA = dataA.length - 1, indexB = dataB.length - 1; indexA >= 0 || indexB >= 0; --indexA, --indexB) {
    const vA = indexA >= 0 ? dataA[indexA] : 0;
    const vB = indexB >= 0 ? dataB[indexB] : 0;
    const current = vA + vB + reminder;
    data.push(current >>> 0);
    reminder = ~~(current / 0x100000000);
  }
  if (reminder !== 0) {
    data.push(reminder);
  }
  return { sign: arrayIntA.sign, data: data.reverse() };
}

/**
 * Add one to a given positive ArrayInt
 * @internal
 */
export function addOneToPositiveArrayInt(arrayInt: ArrayInt): ArrayInt {
  arrayInt.sign = 1; // handling case { sign: -1, data: [0,...,0] }
  const data = arrayInt.data;
  for (let index = data.length - 1; index >= 0; --index) {
    if (data[index] === 0xffffffff) {
      data[index] = 0;
    } else {
      data[index] += 1;
      return arrayInt;
    }
  }
  data.unshift(1);
  return arrayInt;
}

/** @internal */
function isStrictlySmaller(dataA: number[], dataB: number[]): boolean {
  const maxLength = Math.max(dataA.length, dataB.length);
  for (let index = 0; index < maxLength; ++index) {
    const indexA = index + dataA.length - maxLength;
    const indexB = index + dataB.length - maxLength;
    const vA = indexA >= 0 ? dataA[indexA] : 0;
    const vB = indexB >= 0 ? dataB[indexB] : 0;
    if (vA < vB) return true;
    if (vA > vB) return false;
  }
  return false;
}

/**
 * Substract two ArrayInt
 * @internal
 */
export function substractArrayIntToNew(arrayIntA: ArrayInt, arrayIntB: ArrayInt): ArrayInt {
  if (arrayIntA.sign !== arrayIntB.sign) {
    return addArrayIntToNew(arrayIntA, { sign: -arrayIntB.sign as -1 | 1, data: arrayIntB.data });
  }
  const dataA = arrayIntA.data;
  const dataB = arrayIntB.data;
  if (isStrictlySmaller(dataA, dataB)) {
    const out = substractArrayIntToNew(arrayIntB, arrayIntA);
    out.sign = -out.sign as -1 | 1;
    return out;
  }
  const data: number[] = [];
  let reminder = 0;
  for (let indexA = dataA.length - 1, indexB = dataB.length - 1; indexA >= 0 || indexB >= 0; --indexA, --indexB) {
    const vA = indexA >= 0 ? dataA[indexA] : 0;
    const vB = indexB >= 0 ? dataB[indexB] : 0;
    const current = vA - vB - reminder;
    data.push(current >>> 0);
    reminder = current < 0 ? 1 : 0;
  }
  return { sign: arrayIntA.sign, data: data.reverse() };
}

/**
 * Trim uneeded zeros in ArrayInt
 * and uniform notation for zero: {sign: 1, data: [0]}
 */
export function trimArrayIntInplace(arrayInt: ArrayInt) {
  const data = arrayInt.data;
  let firstNonZero = 0;
  for (; firstNonZero !== data.length && data[firstNonZero] === 0; ++firstNonZero) {}
  if (firstNonZero === data.length) {
    // only zeros
    arrayInt.sign = 1;
    arrayInt.data = [0];
    return arrayInt;
  }
  data.splice(0, firstNonZero);
  return arrayInt;
}

// Helpers specific to 64 bits versions

/** @internal */
export type ArrayInt64 = Required<ArrayInt> & { data: [number, number] };

/**
 * We only accept safe integers here
 * @internal
 */
export function fromNumberToArrayInt64(out: ArrayInt64, n: number): ArrayInt64 {
  if (n < 0) {
    const posN = -n;
    out.sign = -1;
    out.data[0] = ~~(posN / 0x100000000);
    out.data[1] = posN >>> 0;
  } else {
    out.sign = 1;
    out.data[0] = ~~(n / 0x100000000);
    out.data[1] = n >>> 0;
  }
  return out;
}

/**
 * Substract two ArrayInt of 64 bits on 64 bits.
 * With arrayIntA - arrayIntB >= 0
 * @internal
 */
export function substractArrayInt64(out: ArrayInt64, arrayIntA: ArrayInt64, arrayIntB: ArrayInt64): ArrayInt64 {
  const lowA = arrayIntA.data[1];
  const highA = arrayIntA.data[0];
  const signA = arrayIntA.sign;
  const lowB = arrayIntB.data[1];
  const highB = arrayIntB.data[0];
  const signB = arrayIntB.sign;

  // Requirement: arrayIntA - arrayIntB >= 0
  out.sign = 1;

  if (signA === 1 && signB === -1) {
    // Operation is a simple sum of arrayIntA + abs(arrayIntB)
    const low = lowA + lowB;
    const high = highA + highB + (low > 0xffffffff ? 1 : 0);
    out.data[0] = high >>> 0;
    out.data[1] = low >>> 0;
    return out;
  }
  // signA === -1 with signB === 1 is impossible given: arrayIntA - arrayIntB >= 0

  // Operation is a substraction
  let lowFirst = lowA;
  let highFirst = highA;
  let lowSecond = lowB;
  let highSecond = highB;
  if (signA === -1) {
    lowFirst = lowB;
    highFirst = highB;
    lowSecond = lowA;
    highSecond = highA;
  }
  let reminderLow = 0;
  let low = lowFirst - lowSecond;
  if (low < 0) {
    reminderLow = 1;
    low = low >>> 0;
  }
  out.data[0] = highFirst - highSecond - reminderLow;
  out.data[1] = low;
  return out;
}
