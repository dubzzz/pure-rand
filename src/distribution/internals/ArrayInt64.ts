import type { ArrayInt } from './ArrayInt';

// Helpers specific to 64 bits versions

/** @internal */
export type ArrayInt64 = Required<ArrayInt> & { data: [number, number] };

/**
 * We only accept safe integers here
 * @internal
 */
export function fromNumberToArrayInt64(out: ArrayInt64, n: number): ArrayInt64 {
  if (n < 0) {
    const posN = -n;
    out.sign = -1;
    out.data[0] = ~~(posN / 0x100000000);
    out.data[1] = posN >>> 0;
  } else {
    out.sign = 1;
    out.data[0] = ~~(n / 0x100000000);
    out.data[1] = n >>> 0;
  }
  return out;
}

/**
 * Substract two ArrayInt of 64 bits on 64 bits.
 * With arrayIntA - arrayIntB >= 0
 * @internal
 */
export function substractArrayInt64(out: ArrayInt64, arrayIntA: ArrayInt64, arrayIntB: ArrayInt64): ArrayInt64 {
  const lowA = arrayIntA.data[1];
  const highA = arrayIntA.data[0];
  const signA = arrayIntA.sign;
  const lowB = arrayIntB.data[1];
  const highB = arrayIntB.data[0];
  const signB = arrayIntB.sign;

  // Requirement: arrayIntA - arrayIntB >= 0
  out.sign = 1;

  if (signA === 1 && signB === -1) {
    // Operation is a simple sum of arrayIntA + abs(arrayIntB)
    const low = lowA + lowB;
    const high = highA + highB + (low > 0xffffffff ? 1 : 0);
    out.data[0] = high >>> 0;
    out.data[1] = low >>> 0;
    return out;
  }
  // signA === -1 with signB === 1 is impossible given: arrayIntA - arrayIntB >= 0

  // Operation is a substraction
  let lowFirst = lowA;
  let highFirst = highA;
  let lowSecond = lowB;
  let highSecond = highB;
  if (signA === -1) {
    lowFirst = lowB;
    highFirst = highB;
    lowSecond = lowA;
    highSecond = highA;
  }
  let reminderLow = 0;
  let low = lowFirst - lowSecond;
  if (low < 0) {
    reminderLow = 1;
    low = low >>> 0;
  }
  out.data[0] = highFirst - highSecond - reminderLow;
  out.data[1] = low;
  return out;
}
