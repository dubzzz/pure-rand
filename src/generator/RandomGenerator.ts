export interface RandomGenerator {
  /** Produce a fully independent clone of the current instance */
  clone(): RandomGenerator;
  /**
   * Generate next random value along with the next generator (does not impact current instance).
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  next(): [number, RandomGenerator];
  /** Compute the jumped generator (does not impact current instance) */
  jump?(): RandomGenerator;
  /**
   * Generate next value BUT alters current generator.
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  unsafeNext(): number;
  /** Jump current generator */
  unsafeJump?(): void;
}

export function unsafeGenerateN(rng: RandomGenerator, num: number): number[] {
  const out: number[] = [];
  for (let idx = 0; idx != num; ++idx) {
    out.push(rng.unsafeNext());
  }
  return out;
}

export function generateN(rng: RandomGenerator, num: number): [number[], RandomGenerator] {
  const nextRng = rng.clone();
  const out = unsafeGenerateN(nextRng, num);
  return [out, nextRng];
}

export function unsafeSkipN(rng: RandomGenerator, num: number): void {
  for (let idx = 0; idx != num; ++idx) {
    rng.unsafeNext();
  }
}

export function skipN(rng: RandomGenerator, num: number): RandomGenerator {
  const nextRng = rng.clone();
  unsafeSkipN(nextRng, num);
  return nextRng;
}
