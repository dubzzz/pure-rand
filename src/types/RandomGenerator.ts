export interface RandomGenerator {
  /** Produce a fully independent clone of the current instance */
  clone(): RandomGenerator;
  /**
   * Generate next random value along with the next generator (does not impact current instance).
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  next(): [number, RandomGenerator];
  /** Compute the jumped generator (does not impact current instance) */
  jump?(): RandomGenerator;
  /**
   * Generate next value BUT alters current generator.
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  unsafeNext(): number;
  /** Jump current generator */
  unsafeJump?(): void;
  /** Access to the internal state of a RandomGenerator in a read-only fashion */
  getState?(): readonly number[];
}
