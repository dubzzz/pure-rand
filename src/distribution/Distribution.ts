import { RandomGenerator } from '../generator/RandomGenerator';

/**
 * Generate random value based on a given RandomGenerator.
 * Return the generated value and an offsetted version of the RandomGenerator (never alters the source rng).
 * @public
 */
export type Distribution<T> = (rng: RandomGenerator) => [T, RandomGenerator];
