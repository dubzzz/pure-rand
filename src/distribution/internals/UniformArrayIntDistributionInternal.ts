import RandomGenerator from '../../generator/RandomGenerator';
import { ArrayInt } from './ArrayInt';
import { uniformIntDistributionInternal } from './UniformIntDistributionInternal';

/**
 * Uniformly generate ArrayInt in range [0 ; rangeSize[
 *
 * @remarks
 * In the worst case scenario it may discard half of the randomly generated value.
 * Worst case being: most significant number is 1 and remaining part evaluates to 0.
 *
 * @internal
 */
export function uniformArrayIntDistributionInternal(
  out: ArrayInt['data'],
  rangeSize: ArrayInt['data'],
  rng: RandomGenerator
): [ArrayInt['data'], RandomGenerator] {
  const rangeLength = rangeSize.length;
  let nrng = rng;

  // We iterate until we find a valid value for arrayInt
  while (true) {
    // We compute a new value for arrayInt
    for (let index = 0; index !== rangeLength; ++index) {
      const indexRangeSize = index === 0 ? rangeSize[0] + 1 : 0x100000000;
      const g = uniformIntDistributionInternal(indexRangeSize, nrng);
      out[index] = g[0];
      nrng = g[1];
    }

    // If in the correct range we can return it
    for (let index = 0; index !== rangeLength; ++index) {
      const current = out[index];
      const currentInRange = rangeSize[index];
      if (current < currentInRange) {
        return [out, nrng]; // arrayInt < rangeSize
      } else if (current > currentInRange) {
        break; // arrayInt > rangeSize
      }
    }
    // Otherwise we need to try another one
  }
}
