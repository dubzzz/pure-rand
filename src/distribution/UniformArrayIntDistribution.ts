import { Distribution } from './Distribution';
import { RandomGenerator } from '../generator/RandomGenerator';
import { ArrayInt } from './internals/ArrayInt';
import { unsafeUniformArrayIntDistribution } from './UnsafeUniformArrayIntDistribution';

/**
 * Uniformly generate random ArrayInt values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 *
 * @public
 */
function uniformArrayIntDistribution(from: ArrayInt, to: ArrayInt): Distribution<ArrayInt>;
/**
 * Uniformly generate random ArrayInt values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 * @param rng - Instance of RandomGenerator to extract random values from
 *
 * @public
 */
function uniformArrayIntDistribution(from: ArrayInt, to: ArrayInt, rng: RandomGenerator): [ArrayInt, RandomGenerator];
function uniformArrayIntDistribution(from: ArrayInt, to: ArrayInt, rng?: RandomGenerator) {
  if (rng != null) {
    const nextRng = rng.clone();
    return [unsafeUniformArrayIntDistribution(from, to, nextRng), nextRng];
  }
  return function (rng: RandomGenerator) {
    const nextRng = rng.clone();
    return [unsafeUniformArrayIntDistribution(from, to, nextRng), nextRng];
  };
}

export { uniformArrayIntDistribution };
