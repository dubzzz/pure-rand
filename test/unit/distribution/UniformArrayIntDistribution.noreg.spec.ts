import fc from 'fast-check';

import { ArrayInt } from '../../../src/distribution/internals/ArrayInt';
import { uniformArrayIntDistribution } from '../../../src/distribution/UniformArrayIntDistribution';
import mersenne from '../../../src/generator/MersenneTwister';
import { RandomGenerator } from '../../../src/generator/RandomGenerator';

describe('uniformArrayIntDistribution [non regression]', () => {
  it.each`
    from                                            | to                                                                     | topic
    ${{ sign: -1, data: [1] }}                      | ${{ sign: 1, data: [1] }}                                              | ${'3 states [-1, 0, 1]'}
    ${{ sign: -1, data: [2] }}                      | ${{ sign: 1, data: [1] }}                                              | ${'4 states [-2, -1, 0, 1]'}
    ${{ sign: -1, data: [0x80000000] }}             | ${{ sign: 1, data: [0x7fffffff] }}                                     | ${'32-bit signed'}
    ${{ sign: 1, data: [0x00000000] }}              | ${{ sign: 1, data: [0xffffffff] }}                                     | ${'32-bit unsigned'}
    ${{ sign: -1, data: [0x80000000, 0x00000000] }} | ${{ sign: 1, data: [0x7fffffff, 0xffffffff] }}                         | ${'64-bit signed'}
    ${{ sign: 1, data: [0x00000000, 0x00000000] }}  | ${{ sign: 1, data: [0xffffffff, 0xffffffff] }}                         | ${'64-bit unsigned'}
    ${{ sign: -1, data: [0x80000000, 0, 0, 0] }}    | ${{ sign: 1, data: [0x7fffffff, 0xffffffff, 0xffffffff, 0xffffffff] }} | ${'128-bit signed'}
    ${{ sign: 1, data: [0x12345678, 0x90abcdef] }}  | ${{ sign: 1, data: [0xfedcba09, 0x87654321] }}                         | ${'fuzzy'}
    ${{ sign: 1, data: [0, 0] }}                    | ${{ sign: 1, data: [0, 5] }}                                           | ${'trailing zeros'}
  `('Should not change its output in asked range except for major bumps ($topic)', ({ from, to }) => {
    // Remark:
    // ========================
    // This test is purely there to ensure that we do not introduce any regression
    // during a commit without noticing it.
    // The values we expect in the output are just a snapshot taken at a certain time
    // in the past. They might be wrong values with bugs.

    let rng: RandomGenerator = mersenne(0);
    const distribution = uniformArrayIntDistribution(from, to);

    const values: ArrayInt[] = [];
    for (let idx = 0; idx !== 10; ++idx) {
      const [v, nrng] = distribution(rng);
      values.push(v);
      rng = nrng;
    }
    expect(values).toMatchSnapshot();
  });

  // Skip next tests if BigInt is not supported
  if (typeof BigInt === 'undefined') return it('no test', () => expect(true).toBe(true));

  it('Should always generate values within the range [from ; to]', () =>
    fc.assert(
      fc.property(fc.integer().noShrink(), arrayIntArb(), arrayIntArb(), (seed, a, b) => {
        const [from, to] = arrayIntToBigInt(a) < arrayIntToBigInt(b) ? [a, b] : [b, a];
        const [v, _nrng] = uniformArrayIntDistribution(from, to)(mersenne(seed));
        const vBigInt = arrayIntToBigInt(v);
        return vBigInt >= arrayIntToBigInt(from) && vBigInt <= arrayIntToBigInt(to);
      })
    ));

  it('Should always trim the zeros from the resulting value', () =>
    fc.assert(
      fc.property(fc.integer().noShrink(), arrayIntArb(), arrayIntArb(), (seed, a, b) => {
        const [from, to] = arrayIntToBigInt(a) < arrayIntToBigInt(b) ? [a, b] : [b, a];
        const [v, _nrng] = uniformArrayIntDistribution(from, to)(mersenne(seed));
        expect(v.data).not.toHaveLength(0);
        if (v.data.length !== 1) {
          expect(v.data[0]).not.toBe(0); // do not start by zero when data has multiple values
        } else if (v.data[0] === 0) {
          expect(v.sign).toBe(1); // zero has sign=1
        }
      })
    ));

  it('Should always produce valid ArrayInt', () =>
    fc.assert(
      fc.property(fc.integer().noShrink(), arrayIntArb(), arrayIntArb(), (seed, a, b) => {
        const [from, to] = arrayIntToBigInt(a) < arrayIntToBigInt(b) ? [a, b] : [b, a];
        const [v, _nrng] = uniformArrayIntDistribution(from, to)(mersenne(seed));
        expect([-1, 1]).toContainEqual(v.sign); // sign is either 1 or -1
        expect(v.data).not.toHaveLength(0); // data is never empty
        for (const d of v.data) {
          // data in [0 ; 0xffffffff]
          expect(d).toBeGreaterThanOrEqual(0);
          expect(d).toBeLessThanOrEqual(0xffffffff);
        }
      })
    ));
});

// Helpers

const arrayIntArb = () =>
  fc.record({
    sign: fc.constantFrom(1 as const, -1 as const),
    data: fc.array(fc.integer({ min: 0, max: 0xffffffff })),
  });

function arrayIntToBigInt(arrayInt: ArrayInt): bigint {
  let current = BigInt(0);
  for (let index = 0; index < arrayInt.data.length; ++index) {
    current <<= BigInt(32);
    current += BigInt(arrayInt.data[index]);
  }
  return current * BigInt(arrayInt.sign);
}
