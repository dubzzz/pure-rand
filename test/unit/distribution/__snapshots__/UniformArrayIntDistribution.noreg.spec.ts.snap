// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (3 states [-1, 0, 1]) 1`] = `
[
  {
    "data": [
      1,
    ],
    "sign": -1,
  },
  {
    "data": [
      0,
    ],
    "sign": 1,
  },
  {
    "data": [
      1,
    ],
    "sign": -1,
  },
  {
    "data": [
      0,
    ],
    "sign": 1,
  },
  {
    "data": [
      0,
    ],
    "sign": 1,
  },
  {
    "data": [
      1,
    ],
    "sign": 1,
  },
  {
    "data": [
      1,
    ],
    "sign": -1,
  },
  {
    "data": [
      1,
    ],
    "sign": 1,
  },
  {
    "data": [
      1,
    ],
    "sign": -1,
  },
  {
    "data": [
      1,
    ],
    "sign": -1,
  },
]
`;

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (4 states [-2, -1, 0, 1]) 1`] = `
[
  {
    "data": [
      1,
    ],
    "sign": -1,
  },
  {
    "data": [
      1,
    ],
    "sign": -1,
  },
  {
    "data": [
      2,
    ],
    "sign": -1,
  },
  {
    "data": [
      0,
    ],
    "sign": 1,
  },
  {
    "data": [
      2,
    ],
    "sign": -1,
  },
  {
    "data": [
      1,
    ],
    "sign": -1,
  },
  {
    "data": [
      2,
    ],
    "sign": -1,
  },
  {
    "data": [
      2,
    ],
    "sign": -1,
  },
  {
    "data": [
      2,
    ],
    "sign": -1,
  },
  {
    "data": [
      1,
    ],
    "sign": 1,
  },
]
`;

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (32-bit signed) 1`] = `
[
  {
    "data": [
      1748719057,
    ],
    "sign": -1,
  },
  {
    "data": [
      1277901399,
    ],
    "sign": 1,
  },
  {
    "data": [
      243580376,
    ],
    "sign": 1,
  },
  {
    "data": [
      1171049868,
    ],
    "sign": 1,
  },
  {
    "data": [
      2051556033,
    ],
    "sign": 1,
  },
  {
    "data": [
      806729177,
    ],
    "sign": -1,
  },
  {
    "data": [
      1686997841,
    ],
    "sign": 1,
  },
  {
    "data": [
      602801999,
    ],
    "sign": 1,
  },
  {
    "data": [
      2034131043,
    ],
    "sign": 1,
  },
  {
    "data": [
      855081807,
    ],
    "sign": -1,
  },
]
`;

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (32-bit unsigned) 1`] = `
[
  {
    "data": [
      398764591,
    ],
    "sign": 1,
  },
  {
    "data": [
      3425385047,
    ],
    "sign": 1,
  },
  {
    "data": [
      2391064024,
    ],
    "sign": 1,
  },
  {
    "data": [
      3318533516,
    ],
    "sign": 1,
  },
  {
    "data": [
      4199039681,
    ],
    "sign": 1,
  },
  {
    "data": [
      1340754471,
    ],
    "sign": 1,
  },
  {
    "data": [
      3834481489,
    ],
    "sign": 1,
  },
  {
    "data": [
      2750285647,
    ],
    "sign": 1,
  },
  {
    "data": [
      4181614691,
    ],
    "sign": 1,
  },
  {
    "data": [
      1292401841,
    ],
    "sign": 1,
  },
]
`;

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (64-bit signed) 1`] = `
[
  {
    "data": [
      1748719056,
      3370736011,
    ],
    "sign": -1,
  },
  {
    "data": [
      1706118332,
      2757602565,
    ],
    "sign": -1,
  },
  {
    "data": [
      1277901399,
      1682652230,
    ],
    "sign": 1,
  },
  {
    "data": [
      156067239,
      976433780,
    ],
    "sign": -1,
  },
  {
    "data": [
      2051556033,
      1252949478,
    ],
    "sign": 1,
  },
  {
    "data": [
      1686997841,
      1827923621,
    ],
    "sign": 1,
  },
  {
    "data": [
      718892300,
      2331500859,
    ],
    "sign": -1,
  },
  {
    "data": [
      602801999,
      1589190063,
    ],
    "sign": 1,
  },
  {
    "data": [
      91833517,
      113352605,
    ],
    "sign": -1,
  },
  {
    "data": [
      855081806,
      165444877,
    ],
    "sign": -1,
  },
]
`;

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (64-bit unsigned) 1`] = `
[
  {
    "data": [
      398764591,
      924231285,
    ],
    "sign": 1,
  },
  {
    "data": [
      441365315,
      1537364731,
    ],
    "sign": 1,
  },
  {
    "data": [
      3425385047,
      1682652230,
    ],
    "sign": 1,
  },
  {
    "data": [
      1991416408,
      3318533516,
    ],
    "sign": 1,
  },
  {
    "data": [
      4199039681,
      1252949478,
    ],
    "sign": 1,
  },
  {
    "data": [
      3834481489,
      1827923621,
    ],
    "sign": 1,
  },
  {
    "data": [
      1428591347,
      1963466437,
    ],
    "sign": 1,
  },
  {
    "data": [
      2750285647,
      1589190063,
    ],
    "sign": 1,
  },
  {
    "data": [
      2055650130,
      4181614691,
    ],
    "sign": 1,
  },
  {
    "data": [
      1292401841,
      4129522419,
    ],
    "sign": 1,
  },
]
`;

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (128-bit signed) 1`] = `
[
  {
    "data": [
      1748719056,
      3370736010,
      2816357183,
      3853601981,
    ],
    "sign": -1,
  },
  {
    "data": [
      156067239,
      976433779,
      500614853,
      95927615,
    ],
    "sign": -1,
  },
  {
    "data": [
      806729176,
      4170864552,
      85997393,
      4002718120,
    ],
    "sign": -1,
  },
  {
    "data": [
      2034131043,
      1284876248,
      1292401841,
      4129522419,
    ],
    "sign": 1,
  },
  {
    "data": [
      1795211326,
      1531785974,
      4134450502,
      2385128833,
    ],
    "sign": -1,
  },
  {
    "data": [
      2053645792,
      1692613941,
      366524171,
      113385321,
    ],
    "sign": -1,
  },
  {
    "data": [
      1853562077,
      1566726015,
      2066782079,
      755308636,
    ],
    "sign": -1,
  },
  {
    "data": [
      1661879776,
      270445703,
      2566893310,
      3446147775,
    ],
    "sign": -1,
  },
  {
    "data": [
      1354754446,
      463129187,
      3709609525,
      3543550860,
    ],
    "sign": 1,
  },
  {
    "data": [
      438279108,
      656229423,
      3044602495,
      580073460,
    ],
    "sign": 1,
  },
]
`;

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (fuzzy) 1`] = `
[
  {
    "data": [
      515072292,
      2825943070,
    ],
    "sign": 1,
  },
  {
    "data": [
      1229651181,
      3905788591,
    ],
    "sign": 1,
  },
  {
    "data": [
      746785211,
      3964543210,
    ],
    "sign": 1,
  },
  {
    "data": [
      498191675,
      3918613334,
    ],
    "sign": 1,
  },
  {
    "data": [
      4272487041,
      2957880514,
    ],
    "sign": 1,
  },
  {
    "data": [
      932030350,
      1930601697,
    ],
    "sign": 1,
  },
  {
    "data": [
      3730804943,
      4109830709,
    ],
    "sign": 1,
  },
  {
    "data": [
      2696483921,
      123627591,
    ],
    "sign": 1,
  },
  {
    "data": [
      3623953413,
      1926563625,
    ],
    "sign": 1,
  },
  {
    "data": [
      1558369374,
      3767932950,
    ],
    "sign": 1,
  },
]
`;

exports[`uniformArrayIntDistribution [non regression] Should not change its output in asked range except for major bumps (trailing zeros) 1`] = `
[
  {
    "data": [
      2,
    ],
    "sign": 1,
  },
  {
    "data": [
      1,
    ],
    "sign": 1,
  },
  {
    "data": [
      5,
    ],
    "sign": 1,
  },
  {
    "data": [
      0,
    ],
    "sign": 1,
  },
  {
    "data": [
      0,
    ],
    "sign": 1,
  },
  {
    "data": [
      4,
    ],
    "sign": 1,
  },
  {
    "data": [
      1,
    ],
    "sign": 1,
  },
  {
    "data": [
      3,
    ],
    "sign": 1,
  },
  {
    "data": [
      0,
    ],
    "sign": 1,
  },
  {
    "data": [
      2,
    ],
    "sign": 1,
  },
]
`;
